src/box/lua/schema.lua

```diff
@@ -3677,30 +3677,101 @@ local function role_check_grant_revoke_of_sys_priv(priv)
     end
 end
 
-box.schema.role.grant = function(user_name, ...)
+local function box_schema_role_grant(user_name, ...)
     local uid = role_resolve(user_name)
     if uid == nil then
         box.error(box.error.NO_SUCH_ROLE, user_name)
     end
     role_check_grant_revoke_of_sys_priv(...)
     return grant(uid, user_name, ...)
 end
-box.schema.role.revoke = function(user_name, ...)
+
+local function box_schema_role_revoke(user_name, ...)
     local uid = role_resolve(user_name)
     if uid == nil then
         box.error(box.error.NO_SUCH_ROLE, user_name)
     end
     role_check_grant_revoke_of_sys_priv(...)
     return revoke(uid, user_name, ...)
 end
-box.schema.role.info = function(role_name)
+
+local function box_schema_role_info(role_name)
     local rid = role_resolve(role_name)
     if rid == nil then
         box.error(box.error.NO_SUCH_ROLE, role_name)
     end
     return info(rid)
 end
 
+-- Wrap a function into transaction if none is active
+local function transactional_wrapper(func)
+    return function(...)
+        local atomic
+
+        if not box.is_in_txn() then
+            atomic = true
+        else
+            atomic = false
+        end
+
+        if atomic then box.begin() end
```

Comment:
Actually there's `box.atomic` function that does exactly what you need, and I believe it is perfectly optimized.

Reply:
Reused it.